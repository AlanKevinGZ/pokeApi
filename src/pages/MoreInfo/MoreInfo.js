import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate, useParams } from "react-router-dom";
import { fetchPokemonByName } from "../../store/PokemonSlice";
import { BackButton, Container, InfoSection, InfoText, LoadingMessage, PokemonCard, PokemonImage, PokemonName, StatBarBackground, StatBarFill, StatItem, StatName, StatsSection, StatsTitle, TypeBadge, TypesContainer} from './style/MoreStyles'
import Header from "../../components/Header/Header";

function MoreInfo() {
  const { id } = useParams();
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const { pokemonDetails, loadingDetails, errorDetails } = useSelector(
    (state) => state.pokemon
  );

  useEffect(() => {
    console.log(id);
    
    if (id) {
      dispatch(fetchPokemonByName(id));
    }

   
  }, [id, dispatch]);

  if (loadingDetails) {
    return (
      <div>
        <Header title={`Buscando ${id}...`} />
        <LoadingMessage>Cargando detalles...</LoadingMessage>
      </div>
    )
  }

   const handleGoBack = () => {
    navigate('/home')
  }

  if (errorDetails) {
    return (
      <div>
        <Header title="Error" />
        <div style={{ padding: '20px' }}>
          <p style={{ color: 'red' }}>Error: {errorDetails}</p>
          <button onClick={handleGoBack}>Volver al inicio</button>
        </div>
      </div>
    )
  }

  if (!pokemonDetails) {
    return (
      <div>
        <Header title="Pokémon no encontrado" />
        <button onClick={handleGoBack}>Volver al inicio</button>
      </div>
    )
  }


  return <div>
     <div>
      <Header title={pokemonDetails.name.toUpperCase()} />
      
      <Container>
        <BackButton onClick={handleGoBack}>
          ← Volver
        </BackButton>

        <PokemonCard>
          {/* Imagen del pokémon */}
          <PokemonImage 
            src={pokemonDetails.sprites.front_default} 
            alt={pokemonDetails.name}
          />

          {/* Nombre */}
          <PokemonName>{pokemonDetails.name}</PokemonName>

          {/* Información básica */}
          <InfoSection>
            <InfoText>
              <strong>ID:</strong> #{pokemonDetails.id}
            </InfoText>
            <InfoText>
              <strong>Altura:</strong> {pokemonDetails.height / 10} m
            </InfoText>
            <InfoText>
              <strong>Peso:</strong> {pokemonDetails.weight / 10} kg
            </InfoText>
            
            {/* Tipos */}
            <InfoText>
              <strong>Tipo:</strong>
              <TypesContainer>
                {pokemonDetails.types.map((t, index) => (
                  <TypeBadge 
                    key={t.type.name}
  
                    first={index === 0}
                  >
                    {t.type.name}
                  </TypeBadge>
                ))}
              </TypesContainer>
            </InfoText>

            {/* Habilidades */}
            <InfoText>
              <strong>Habilidades:</strong>{' '}
              {pokemonDetails.abilities.map(a => a.ability.name).join(', ')}
            </InfoText>

            {/* Estadísticas */}
            <StatsSection>
              <StatsTitle>Estadísticas:</StatsTitle>
              {pokemonDetails.stats.map((stat) => (
                <StatItem key={stat.stat.name}>
                  <StatName>{stat.stat.name}:</StatName> {stat.base_stat}
                </StatItem>
              ))}
            </StatsSection>
          </InfoSection>
        </PokemonCard>
      </Container>
    </div>

  </div>;
}

export default MoreInfo;
