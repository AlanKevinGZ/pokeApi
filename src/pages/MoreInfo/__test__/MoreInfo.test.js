// src/pages/MoreInfo/__test__/MoreInfo.test.js
import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import "@testing-library/jest-dom";
import MoreInfo from "../MoreInfo";
import pokemonReducer from "../../../store/PokemonSlice";

// ========================================
// MOCK THEME
// ========================================
const mockTheme = {
  colors: {
    buttons: '#3498db',
    primary: '#2c3e50',
    secondary: '#ecf0f1',
    text: '#ffffff',
    background: '#f5f5f5',
  },
};

// ========================================
// MOCK useNavigate
// ========================================
const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

// ========================================
// MOCK Header Component
// ========================================
jest.mock("../../../components/Header/Header", () => {
  return function MockHeader({ title }) {
    return <div data-testid="header">{title}</div>;
  };
});

// ========================================
// MOCK fetch global
// ========================================
global.fetch = jest.fn();

// ========================================
// MOCK POKEMON DATA
// ========================================
const mockPokemonData = {
  id: 25,
  name: "pikachu",
  height: 4,
  weight: 60,
  sprites: {
    front_default: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png"
  },
  types: [
    {
      type: {
        name: "electric"
      }
    }
  ],
  abilities: [
    {
      ability: {
        name: "static"
      }
    },
    {
      ability: {
        name: "lightning-rod"
      }
    }
  ],
  stats: [
    {
      base_stat: 35,
      stat: {
        name: "hp"
      }
    },
    {
      base_stat: 55,
      stat: {
        name: "attack"
      }
    },
    {
      base_stat: 40,
      stat: {
        name: "defense"
      }
    }
  ]
};

// ========================================
// HELPER: Render with all providers
// ========================================
const renderWithProviders = (initialState = {}, route = "/more/pikachu") => {
  const defaultState = {
    list: [],
    selected: null,
    loading: false,
    error: null,
    pokemonDetails: null,
    loadingDetails: false,
    errorDetails: null,
    ...initialState
  };

  const store = configureStore({
    reducer: {
      pokemon: pokemonReducer,
    },
    preloadedState: {
      pokemon: defaultState
    }
  });

  return {
    ...render(
      <Provider store={store}>
        <ThemeProvider theme={mockTheme}>
          <MemoryRouter initialEntries={[route]}>
            <Routes>
              <Route path="/more/:id" element={<MoreInfo />} />
            </Routes>
          </MemoryRouter>
        </ThemeProvider>
      </Provider>
    ),
    store
  };
};

// ========================================
// TESTS
// ========================================
describe("MoreInfo Component", () => {

  beforeEach(() => {
    jest.clearAllMocks();
    mockNavigate.mockClear();
    global.fetch.mockResolvedValue({
      json: async () => mockPokemonData,
    });
  });

  // ========================================
  // TEST 1: Shows loading message
  // ========================================
  test("shows loading message when loading details", () => {
    renderWithProviders({ loadingDetails: true });
    
    expect(screen.getByText(/cargando detalles/i)).toBeInTheDocument();
  });

 

  // ========================================
  // TEST 3: Shows back button on error
  // ========================================
  test("shows back button on error and navigates when clicked", async () => {
    global.fetch.mockRejectedValueOnce(new Error("Network error"));
    
    renderWithProviders();
    
    await waitFor(() => {
      expect(screen.getByText(/volver al inicio/i)).toBeInTheDocument();
    });

    const backButton = screen.getByText(/volver al inicio/i);
    fireEvent.click(backButton);

    expect(mockNavigate).toHaveBeenCalledWith('/home');
  });


  // ========================================
  // TEST 5: Renders pokemon details
  // ========================================
  test("renders pokemon details correctly", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => mockPokemonData,
    });

    renderWithProviders();

    // Wait for pokemon name to appear
    await waitFor(() => {
      expect(screen.getByText(/pikachu/i)).toBeInTheDocument();
    });

    // Verify ID
    expect(screen.getByText(/#25/i)).toBeInTheDocument();
    
    // Verify height (4 / 10 = 0.4m)
    expect(screen.getByText(/0\.4 m/i)).toBeInTheDocument();
    
    // Verify weight (60 / 10 = 6kg)
    expect(screen.getByText(/6 kg/i)).toBeInTheDocument();
  });

  // ========================================
  // TEST 6: Renders pokemon types
  // ========================================
  test("renders pokemon types", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => mockPokemonData,
    });

    renderWithProviders();

    await waitFor(() => {
      expect(screen.getByText(/electric/i)).toBeInTheDocument();
    });
  });

  // ========================================
  // TEST 7: Renders pokemon abilities
  // ========================================
  test("renders pokemon abilities", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => mockPokemonData,
    });

    renderWithProviders();

    await waitFor(() => {
      expect(screen.getByText(/static, lightning-rod/i)).toBeInTheDocument();
    });
  });

  // ========================================
  // TEST 8: Renders pokemon stats
  // ========================================
  test("renders pokemon stats", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => mockPokemonData,
    });

    renderWithProviders();

    await waitFor(() => {
      expect(screen.getByText(/estadísticas/i)).toBeInTheDocument();
    });

    expect(screen.getByText(/hp:/i)).toBeInTheDocument();
    expect(screen.getByText(/attack:/i)).toBeInTheDocument();
    expect(screen.getByText(/defense:/i)).toBeInTheDocument();
  });


  test("renders pokemon image", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => mockPokemonData,
    });

    renderWithProviders();

    await waitFor(() => {
      const image = screen.getByAltText(/pikachu/i);
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute('src', mockPokemonData.sprites.front_default);
    });
  });


  test("back button navigates to home", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => mockPokemonData,
    });

    renderWithProviders();

    await waitFor(() => {
      expect(screen.getByText(/← volver/i)).toBeInTheDocument();
    });

    const backButton = screen.getByText(/← volver/i);
    fireEvent.click(backButton);

    expect(mockNavigate).toHaveBeenCalledWith('/home');
  });

  test("fetches pokemon by ID from URL params", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => ({
        ...mockPokemonData,
        name: "charmander",
        id: 4
      }),
    });

    renderWithProviders({}, "/more/charmander");

    await waitFor(() => {
      expect(screen.getByText(/charmander/i)).toBeInTheDocument();
    });

    expect(screen.getByText(/#4/i)).toBeInTheDocument();
  });

  
  test("renders header with pokemon name in uppercase", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => mockPokemonData,
    });

    renderWithProviders();

    await waitFor(() => {
      const header = screen.getByTestId("header");
      expect(header).toHaveTextContent("PIKACHU");
    });
  });

 
  test("renders multiple types correctly", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => ({
        ...mockPokemonData,
        types: [
          { type: { name: "grass" } },
          { type: { name: "poison" } }
        ]
      }),
    });

    renderWithProviders();

    await waitFor(() => {
      expect(screen.getByText(/grass/i)).toBeInTheDocument();
    });

    expect(screen.getByText(/poison/i)).toBeInTheDocument();
  });


  test("matches snapshot", async () => {
    global.fetch.mockResolvedValueOnce({
      json: async () => mockPokemonData,
    });

    const { container } = renderWithProviders();

    await waitFor(() => {
      expect(screen.getByText(/pikachu/i)).toBeInTheDocument();
    });

    expect(container).toMatchSnapshot();
  });
});