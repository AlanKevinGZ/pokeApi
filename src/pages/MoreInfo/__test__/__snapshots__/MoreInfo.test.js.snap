// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreInfo Component matches snapshot 1`] = `
<div>
  <div>
    <div>
      <div
        data-testid="header"
      >
        PIKACHU
      </div>
      <div
        class="sc-bRKDuR jnvOHb"
      >
        <button
          class="sc-hvigdm eduGjA"
        >
          ← Volver
        </button>
        <div
          class="sc-fhHczv dQnHSv"
        >
          <img
            alt="pikachu"
            class="sc-ggWZvA gPXyqc"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png"
          />
          <h2
            class="sc-dTvVRJ scUFP"
          >
            pikachu
          </h2>
          <div
            class="sc-jwTyAe gejmtu"
          >
            <p
              class="sc-hjsuWn iOSaGa"
            >
              <strong>
                ID:
              </strong>
               #
              25
            </p>
            <p
              class="sc-hjsuWn iOSaGa"
            >
              <strong>
                Altura:
              </strong>
               
              0.4
               m
            </p>
            <p
              class="sc-hjsuWn iOSaGa"
            >
              <strong>
                Peso:
              </strong>
               
              6
               kg
            </p>
            <p
              class="sc-hjsuWn iOSaGa"
            >
              <strong>
                Tipo:
              </strong>
              <div
                class="sc-dFVmKS ctmSWd"
              >
                <span
                  class="sc-jJLAfE hGEkMP"
                >
                  electric
                </span>
              </div>
            </p>
            <p
              class="sc-hjsuWn iOSaGa"
            >
              <strong>
                Habilidades:
              </strong>
               
              static, lightning-rod
            </p>
            <div
              class="sc-hwkwBN juUUot"
            >
              <h3
                class="sc-kNOymR kbNEyz"
              >
                Estadísticas:
              </h3>
              <div
                class="sc-dYwGCk gHTkRk"
              >
                <strong
                  class="sc-lgpSej fSTDxo"
                >
                  hp
                  :
                </strong>
                 
                35
              </div>
              <div
                class="sc-dYwGCk gHTkRk"
              >
                <strong
                  class="sc-lgpSej fSTDxo"
                >
                  attack
                  :
                </strong>
                 
                55
              </div>
              <div
                class="sc-dYwGCk gHTkRk"
              >
                <strong
                  class="sc-lgpSej fSTDxo"
                >
                  defense
                  :
                </strong>
                 
                40
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
